File: leetcode/cpp/p60-permutation-sequence.cpp
// LeetCode 60. Permutation Sequence.
#include <algorithm>
#include <cassert>
#include <iostream>
#include <numeric>
#include <vector>

const int MAX_N = 9;
int factorial(int n)
{
    assert(n >= 0);
    assert(n <= MAX_N);
    static int facts[MAX_N + 1] = {
        0,
        1,
        2,
        6,
        24,
        120,
        720,
        5040,
        40320,
        362880
    };
    return facts[static_cast<size_t>(n)];
}

class Solution {
public:
    std::string getPermutation(int n, int k)
    {
        assert(n >= 1 && n <= MAX_N);
        assert(k >= 1 && k <= 362880);
        k -= 1;
        std::string s(static_cast<size_t>(n), ' ');
        std::iota(s.begin(), s.end(), '1');
        std::string ans;
        for (int i = n - 1; i > 0; --i) {
            const int div = factorial(i);
            const int idx = k / div;
            k = k % div;
            ans += s[static_cast<size_t>(idx)];
            s.erase(static_cast<size_t>(idx), 1);
        }
        ans += s[0];
        return ans;
    }
};

int main()
{
    struct TestCase {
        int n;
        int k;
        std::string exp;
    };
    TestCase test_cases[] = {
        { 1, 1, "1" },
        { 2, 1, "12" },
        { 2, 2, "21" },
        { 3, 1, "123" },
        { 3, 2, "132" },
        { 3, 3, "213" },
        { 3, 4, "231" },
        { 3, 5, "312" },
        { 3, 6, "321" },
        { 4, 1, "1234" },
        { 4, 2, "1243" },
        { 4, 9, "2314" },
        { 4, 24, "4321" },
        { 5, 1, "12345" },
        { 5, 120, "54321" },
        { 6, 1, "123456" },
        { 6, 720, "654321" },
        { 7, 1, "1234567" },
        { 7, 5040, "7654321" },
        { 8, 1, "12345678" },
        { 8, 40320, "87654321" },
        { 9, 1, "123456789" },
        { 9, 2, "123456798" },
        { 9, 362880, "987654321" },
    };

    Solution solution;
    for (const auto& tc : test_cases) {
        const auto act = solution.getPermutation(tc.n, tc.k);
        if (tc.exp != act) {
            std::cerr << "fail, n: " << tc.n
                      << ", k: " << tc.k
                      << ", exp: " << tc.exp
                      << ", act: " << act
                      << "\n";
        }
    }
}

File: leetcode/swift/p60-permutation-sequence.swift
// LeetCode 60. Permutation Sequence.

func factorial(_ n: Int) -> Int
{
    var fact = 1
    for i in 1...n {
        fact = fact * i
    }
    return fact
}

class Solution {
    func getPermutation(_ n: Int, _ k: Int) -> String {
        var k_ = k - 1
        var a = Array("123456789")
        var ans = String()
        for i in (1 ..< n).reversed() {
            let div = factorial(i)
            let idx = k_ / div
            k_ = k_ % div
            ans.append(a.remove(at:idx))
        }
        ans.append(a[0])
        return ans
    }
}

struct TestCase {
    var n: Int
    var k: Int
    var exp: String
}

let test_cases = [
        TestCase(n: 1, k: 1, exp: "1" ),
        TestCase(n: 2, k: 1, exp: "12" ),
        TestCase(n: 2, k: 2, exp: "21" ),
        TestCase(n: 3, k: 1, exp: "123" ),
        TestCase(n: 3, k: 2, exp: "132" ),
        TestCase(n: 3, k: 3, exp: "213" ),
        TestCase(n: 3, k: 4, exp: "231" ),
        TestCase(n: 3, k: 5, exp: "312" ),
        TestCase(n: 3, k: 6, exp: "321" ),
        TestCase(n: 4, k: 1, exp: "1234" ),
        TestCase(n: 4, k: 2, exp: "1243" ),
        TestCase(n: 4, k: 9, exp: "2314" ),
        TestCase(n: 4, k: 24, exp: "4321" ),
        TestCase(n: 5, k: 1, exp: "12345" ),
        TestCase(n: 5, k: 120, exp: "54321" ),
        TestCase(n: 6, k: 1, exp: "123456" ),
        TestCase(n: 6, k: 720, exp: "654321" ),
        TestCase(n: 7, k: 1, exp: "1234567" ),
        TestCase(n: 7, k: 5040, exp: "7654321" ),
        TestCase(n: 8, k: 1, exp: "12345678" ),
        TestCase(n: 8, k: 40320, exp: "87654321" ),
        TestCase(n: 9, k: 1, exp: "123456789" ),
        TestCase(n: 9, k: 2, exp: "123456798" ),
        TestCase(n: 9, k: 362880, exp: "987654321" ),
]

let solution = Solution()
for tc in test_cases {
    let act = solution.getPermutation(tc.n, tc.k)
    if tc.exp != act {
        print("fail, n: \(tc.n), k: \(tc.k), exp: \(tc.exp), act: \(act)")
    }
}